File: hifiKeyWordSearch.js
// 引入我们需要的两个小伙伴：Express 和 Playwright
const express = require('express');
const { chromium } = require('playwright');

// --- 配置区 ---
// 你的 HiFiNi 账号信息，要好好保管哦~
const USERNAME = "zhuzhuzhuzhu";
const PASSWORD = "CHAOfan0920";

// --- HTML / CSS / JavaScript 前端代码 ---
// 这部分和之前一模一样，我们只是把它从 Python 的世界搬到了 JavaScript 的怀抱里
const HTML_CONTENT = `
<!DOCTYPE html>
<html lang="zh-CN">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>HiFiNi 音乐抓取器 (JS版)</title>
    <script src="https://cdn.tailwindcss.com"></script>
    <style>
        .loader {
            border: 5px solid #f3f3f3;
            border-top: 5px solid #5865f2;
            border-radius: 50%;
            width: 50px;
            height: 50px;
            animation: spin 1s linear infinite;
        }
        @keyframes spin {
            0% { transform: rotate(0deg); }
            100% { transform: rotate(360deg); }
        }
    </style>
</head>
<body class="bg-gray-100 text-gray-800 font-sans">
    <div class="container mx-auto p-4 sm:p-6 md:p-8 max-w-2xl">
        <div class="bg-white p-6 rounded-xl shadow-lg">
            <h1 class="text-2xl sm:text-3xl font-bold text-center mb-2 text-indigo-600">HiFiNi 音乐小助手 🎵 (JS Ver.)</h1>
            <p class="text-center text-gray-500 mb-6">输入想听的歌手或歌名，剩下的交给我吧~</p>
            
            <div class="flex flex-col sm:flex-row gap-2">
                <input type="text" id="keyword" placeholder="例如: 周杰伦" class="flex-grow w-full px-4 py-2 text-lg border-2 border-gray-200 rounded-lg focus:ring-2 focus:ring-indigo-400 focus:border-indigo-400 outline-none transition">
                <button id="search-btn" class="w-full sm:w-auto bg-indigo-500 hover:bg-indigo-600 text-white font-bold py-2 px-6 rounded-lg text-lg transition-transform transform hover:scale-105 focus:outline-none focus:ring-2 focus:ring-indigo-400 focus:ring-opacity-75">
                    <span>开始搜索</span>
                </button>
            </div>
        </div>

        <div id="status-container" class="mt-6 text-center">
            <div id="loader" class="loader mx-auto hidden"></div>
            <p id="status-text" class="text-gray-600 mt-4 text-lg"></p>
        </div>

        <div id="results-container" class="mt-6">
             <ul id="results-list" class="space-y-3">
                <!-- 搜索结果会在这里显示 -->
             </ul>
        </div>
    </div>

    <script>
        const searchBtn = document.getElementById('search-btn');
        const keywordInput = document.getElementById('keyword');
        const loader = document.getElementById('loader');
        const statusText = document.getElementById('status-text');
        const resultsList = document.getElementById('results-list');

        keywordInput.addEventListener('keypress', (event) => {
            if (event.key === 'Enter') {
                searchBtn.click();
            }
        });

        searchBtn.addEventListener('click', async () => {
            const keyword = keywordInput.value.trim();
            if (!keyword) {
                statusText.textContent = '哎呀，关键词不能为空哦~';
                return;
            }

            loader.classList.remove('hidden');
            statusText.textContent = '收到指令！正在启动浏览器登录，请稍等片刻...';
            resultsList.innerHTML = '';
            searchBtn.disabled = true;
            searchBtn.classList.add('opacity-50', 'cursor-not-allowed');
            searchBtn.querySelector('span').textContent = '搜索中...';

            try {
                const response = await fetch(\`/api/search?keyword=\${encodeURIComponent(keyword)}\`);
                
                if (!response.ok) {
                    const errorData = await response.json();
                    throw new Error(errorData.detail || '服务器好像开小差了...');
                }

                const data = await response.json();
                
                loader.classList.add('hidden');
                if (data.length === 0) {
                    statusText.textContent = '好可惜，没有找到相关的歌曲呢... 换个词试试？';
                } else {
                    statusText.textContent = \`太棒了！为你找到了 \${data.length} 首歌曲~\`;
                    data.forEach(song => {
                        const li = document.createElement('li');
                        li.className = 'bg-white p-4 rounded-lg shadow-md flex items-center justify-between transition-transform transform hover:scale-102';
                        
                        const titleSpan = document.createElement('span');
                        titleSpan.textContent = song.title;
                        titleSpan.className = 'flex-grow mr-4 text-gray-700';

                        const downloadLink = document.createElement('a');
                        downloadLink.href = song.url;
                        downloadLink.textContent = '下载';
                        downloadLink.setAttribute('download', song.title.replace(/ /g, '_') + '.mp3');
                        downloadLink.className = 'bg-green-500 hover:bg-green-600 text-white font-bold py-1 px-4 rounded-full text-sm transition';
                        
                        li.appendChild(titleSpan);
                        li.appendChild(downloadLink);
                        resultsList.appendChild(li);
                    });
                }

            } catch (error) {
                loader.classList.add('hidden');
                statusText.textContent = \`出错了: \${error.message}\`;
            } finally {
                searchBtn.disabled = false;
                searchBtn.classList.remove('opacity-50', 'cursor-not-allowed');
                searchBtn.querySelector('span').textContent = '开始搜索';
            }
        });
    </script>
</body>
</html>
`;

// --- Express 后端应用 ---
const app = express();
const PORT = 8000;
let browserContext = null;

async function scrapeHifini(page, keyword) {
    const allMusicData = [];
    const processedUrls = new Set();
    const MAX_PAGES = 3; // 和Python脚本保持一致，最多翻3页

    // 步骤 1: 登录 (如果需要)
    try {
        console.log("检查登录状态...");
        await page.goto("https://hifiti.com/", { timeout: 30000, waitUntil: 'domcontentloaded' });
        const loginButton = page.locator('a[href*="user-login.htm"]');
        if (await loginButton.count() > 0) {
            console.log("尚未登录，现在开始执行登录流程...");
            await page.goto("https://hifiti.com/user-login.htm", { timeout: 20000, waitUntil: 'domcontentloaded' });
            const usernameInput = page.getByPlaceholder("请输入用户名");
            await usernameInput.waitFor({ state: 'visible', timeout: 30000 });
            await usernameInput.fill(USERNAME);
            await page.getByPlaceholder("请输入密码").fill(PASSWORD);
            await page.getByRole("button", { name: "登录" }).click();
            await page.waitForURL("https://hifiti.com/", { timeout: 20000 });
            console.log("登录成功！🎉");
        } else {
            console.log("检测到已登录，跳过登录步骤。");
        }
    } catch (e) {
        console.error(`登录或检查登录状态时出错: ${e}`);
        throw new Error(`登录失败: ${e.message}`);
    }

    // 步骤 2: 搜索关键词
    console.log(`开始搜索关键词 '${keyword}'...`);
    try {
        await page.locator("#searchKeyword").fill(keyword);
        await page.locator(".btn.btn-primary").click();
        await page.waitForLoadState("domcontentloaded");
    } catch (e) {
        console.error(`搜索失败: ${e}`);
        throw new Error(`搜索失败: ${e.message}`);
    }

    // 步骤 3: 循环翻页并采集帖子链接
    let currentPage = 1;
    const collectedLinks = [];

    while (currentPage <= MAX_PAGES) {
        console.log(`--- 开始处理第 ${currentPage} 页 ---`);
        await page.waitForLoadState("networkidle");

        const threadElements = await page.locator("li.media.thread").all();
        console.log(`在第 ${currentPage} 页找到 ${threadElements.length} 个帖子，开始过滤...`);

        for (const item of threadElements) {
            const isHelpPost = await item.locator("a[href='forum-7.htm']").count() > 0;
            if (isHelpPost) {
                const titleElement = await item.locator("div.subject.break-all > a").first();
                const title = await titleElement.innerText();
                console.log(`  - [跳过] 发现互助帖: ${title.trim()}`);
                continue;
            }

            const linkElement = await item.locator("div.subject.break-all > a").first();
            if (linkElement) {
                const title = (await linkElement.innerText()).trim();
                const href = await linkElement.getAttribute("href");
                if (href) {
                    const fullUrl = `https://hifiti.com/${href}`;
                    if (!processedUrls.has(fullUrl)) {
                        collectedLinks.push({ title: title, post_url: fullUrl });
                        processedUrls.add(fullUrl);
                        console.log(`  + [收录] ${title}`);
                    }
                }
            }
        }

        const nextPageButton = page.locator("a.page-link:has-text('▶')");
        if (await nextPageButton.count() > 0) {
            console.log("找到'下一页'按钮，准备翻页...");
            await nextPageButton.click();
            currentPage++;
            await page.waitForLoadState("networkidle", { timeout: 30000 });
        } else {
            console.log("未找到'下一页'按钮，抓取结束。");
            break;
        }
    }
    
    console.log(`\n总共收集到 ${collectedLinks.length} 个有效帖子链接，准备抓取下载地址...`);

    // 步骤 4 & 5: 进入帖子抓取媒体链接
    for (const [index, songInfo] of collectedLinks.entries()) {
        console.log(`--- 处理第 ${index + 1}/${collectedLinks.length} 首：${songInfo.title} ---`);
        let requestListener = null;
        try {
            const mediaUrlPromise = new Promise((resolve, reject) => {
                const timer = setTimeout(() => {
                    page.removeListener('request', requestListener);
                    reject(new Error('超时未抓取到媒体链接'));
                }, 20000);

                requestListener = (request) => {
                    const url = request.url();
                    if (url.match(/\.(mp3|flac|wav)(\?.*)?$/i) || request.resourceType() === 'media') {
                        console.log(`🎶 抓取成功 -> ${url}`);
                        clearTimeout(timer);
                        page.removeListener('request', requestListener);
                        resolve(url);
                    }
                };
                page.on('request', requestListener);
            });

            await page.goto(songInfo.post_url, { waitUntil: "domcontentloaded", timeout: 60000 });
            
            const playButton = page.locator(".aplayer-button, .aplayer-dplayer").first();
            if (await playButton.isVisible()) {
                await playButton.click({ timeout: 5000 }).catch(() => console.log("点击播放按钮失败，但可能已触发请求。"));
            }

            const finalUrl = await mediaUrlPromise;
            allMusicData.push({ title: songInfo.title, url: finalUrl });

        } catch (e) {
            console.error(`❌ 处理帖子 ${songInfo.post_url} 时出错: ${e.message}`);
        } finally {
            if (requestListener) {
                page.removeListener('request', requestListener);
            }
        }
    }
    return allMusicData;
}

// API 路由，负责处理搜索请求
app.get('/api/search', async (req, res) => {
    const keyword = req.query.keyword;
    if (!browserContext) {
        return res.status(503).json({ detail: "浏览器服务尚未准备好，请稍后再试。" });
    }
    if (!keyword) {
        return res.status(400).json({ detail: "关键词不能为空哦！" });
    }
    
    const page = await browserContext.newPage();
    try {
        const results = await scrapeHifini(page, keyword);
        res.json(results);
    } catch (error) {
        res.status(500).json({ detail: error.message || '发生了未知错误' });
    } finally {
        await page.close(); // 每次请求完成后关闭页面，保持环境干净
    }
});

// 根路由，提供我们的HTML页面
app.get('/', (req, res) => {
    res.send(HTML_CONTENT);
});

// --- 启动器 ---
// 使用一个立即执行的异步函数来初始化浏览器并启动服务器
(async () => {
    console.log("应用启动中，正在初始化Playwright...");
    const browser = await chromium.launch({ headless: true }); // headless: true 表示在后台运行
    browserContext = await browser.newContext();
    console.log("Playwright初始化完成！");

    app.listen(PORT, () => {
        console.log("="*50);
        console.log(`🎵 HiFiNi 音乐小助手 (JS版) 已启动！`);
        console.log(`请在浏览器中打开 http://127.0.0.1:${PORT} 访问我哦~`);
        console.log("="*50);
    });
})();

File: hifiKeyWordSearch.py
import time
import json
import sys
from playwright.sync_api import sync_playwright, TimeoutError as PlaywrightTimeoutError

USERNAME = "zhuzhuzhuzhu"
PASSWORD = "CHAOfan0920"
if len(sys.argv) > 1:
    SEARCH_KEYWORD = sys.argv[1]
else:
    SEARCH_KEYWORD = input("请输入您要搜索的关键词: ")
BASE_URL = "https://hifiti.com"
MAX_PAGES = 3  # 设置最大翻页数
QUICK_TEST_MODE = True  # 快速测试模式：True表示只抓10个，False表示按MAX_PAGES抓取
QUICK_TEST_LIMIT = 5   # 快速测试模式下的抓取数量

# 在这里添加不希望抓取的板块链接，例如 'forum-7.htm' 是互助区
EXCLUDED_FORUMS = [
    "forum-7.htm"  # 互助区
]

results = []
processed_urls = set() # 用来存储已经处理过的帖子URL，防止重复抓取

def run():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False)
        context = browser.new_context()
        page = context.new_page()

        # 步骤 1: 登录
        try:
            print("正在尝试登录...")
            page.goto(f"{BASE_URL}/user-login.htm", timeout=60000)
            page.fill("input[name='email']", USERNAME)
            page.fill("input[name='password']", PASSWORD)
            page.click("button:has-text('登录')")
            page.wait_for_url(f"{BASE_URL}/", timeout=30000)
            print("登录成功，已跳转到主页。")
        except PlaywrightTimeoutError:
            print("登录超时或失败，请检查网络或账号密码。")
            browser.close()
            return
        except Exception as e:
            print(f"登录过程中发生未知错误: {e}")
            browser.close()
            return

        # 步骤 2: 搜索
        try:
            print(f"开始在主页搜索关键词: {SEARCH_KEYWORD}")
            page.fill("#searchKeyword", SEARCH_KEYWORD)
            page.click("button.btn.btn-primary")
            page.wait_for_load_state("networkidle", timeout=30000)
        except Exception as e:
            print(f"搜索失败: {e}")
            browser.close()
            return

        # 步骤 3: 循环翻页并采集帖子链接
        collected_links = []
        current_page = 1
        while current_page <= MAX_PAGES:
            print(f"--- 开始处理第 {current_page} 页 ---")
            page.wait_for_load_state("networkidle")
            
            # 获取当前页面的所有帖子
            thread_elements = page.query_selector_all("li.media.thread")
            if not thread_elements:
                print("当前页面没有找到帖子列表。")
                break

            print(f"在第 {current_page} 页找到 {len(thread_elements)} 个帖子，开始过滤...")
            for item in thread_elements:
                # 检查帖子是否属于任何一个被排除的板块
                excluded = False
                for forum_href in EXCLUDED_FORUMS:
                    if item.query_selector(f"a[href='{forum_href}']"):
                        title_element = item.query_selector("div.subject.break-all > a")
                        title = title_element.inner_text() if title_element else "未知标题"
                        print(f"  - [跳过] 发现被排除板块的帖子: {title.strip()}")
                        excluded = True
                        break
                if excluded:
                    continue
                
                link_element = item.query_selector("div.subject.break-all > a")
                if link_element:
                    title = link_element.inner_text().strip()
                    href = link_element.get_attribute("href")
                    full_url = f"{BASE_URL}/{href}"
                    if full_url not in processed_urls:
                        collected_links.append({"title": title, "url": full_url})
                        processed_urls.add(full_url)
                        print(f"  + [收录] {title}")
                        # 如果是快速测试模式，检查是否已达到数量限制
                        if QUICK_TEST_MODE and len(collected_links) >= QUICK_TEST_LIMIT:
                            print(f"\n快速测试模式已启动，并达到 {QUICK_TEST_LIMIT} 个链接的抓取上限，停止采集。")
                            break # 跳出当前页的循环

            # 再次检查是否因为快速模式而需要跳出外层循环
            if QUICK_TEST_MODE and len(collected_links) >= QUICK_TEST_LIMIT:
                break # 跳出翻页循环

            # 寻找并点击“下一页”
            try:
                next_page_button = page.locator("a.page-link:has-text('▶')")
                if next_page_button.count() > 0:
                    print("找到'下一页'按钮，准备翻页...")
                    next_page_button.click()
                    current_page += 1
                    # 等待页面跳转完成
                    page.wait_for_load_state("networkidle", timeout=30000)
                else:
                    print("未找到'下一页'按钮，抓取结束。")
                    break
            except Exception as e:
                print(f"翻页时出错: {e}")
                break
        
        print(f"\n总共收集到 {len(collected_links)} 个有效帖子链接，准备抓取下载地址...")

        # 步骤 4 & 5: 进入帖子抓取媒体链接
        for link_info in collected_links:
            print(f"--- 正在处理: {link_info['title']} ---")
            try:
                page.goto(link_info['url'], timeout=60000, wait_until="domcontentloaded")
                
                # 使用 expect_response 来捕获媒体文件请求
                with page.expect_response(lambda r: r.request.resource_type == "media" or ".mp3" in r.url or ".flac" in r.url, timeout=20000) as response_info:
                    # 尝试点击播放按钮，如果按钮不存在或不可见，可能会超时，但我们依然能通过 expect_response 捕获到链接
                    play_button = page.locator(".aplayer-button, .aplayer-dplayer").first
                    if play_button.is_visible():
                         play_button.click(timeout=5000)
                
                media_response = response_info.value
                media_url = media_response.url
                print(f"  ✅ 抓取成功: {media_url}")
                results.append({"title": link_info['title'], "url": media_url})

            except PlaywrightTimeoutError:
                print(f"  ❌ 处理超时或未找到媒体链接: {link_info['title']}")
            except Exception as e:
                print(f"  ❌ 处理时发生未知错误: {link_info['title']} - {e}")
            
            time.sleep(1) # 每次抓取后稍作停留

        browser.close()

        # 步骤 6: 保存到文件
        with open("music_links.json", "w", encoding="utf-8") as f:
            json.dump(results, f, ensure_ascii=False, indent=4)
        
        print(f"\n🎉 全部完成！成功抓取 {len(results)} 首歌曲，已保存到 music_links.json")

if __name__ == "__main__":
    run()
